{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Function is defined\n",
    "def Slice_Sequence(path_now, window, step):\n",
    "    import sys, os\n",
    "    \n",
    "    \n",
    "    #Starts very handy counters\n",
    "    n1 = 0  #Helps to define the amount of groups we will be working with\n",
    "    n2 = 0  #Keeps track of the group that we are currently working on\n",
    "    independently = 0  #keeps track of whether we are coming from Extract_IDRs or working independently\n",
    "    counter = 0  #keeps track of how many fasta files are present within current group (either .fasta or .txt)\n",
    "    folder_created = 0  #keeps track of whether a folder was already created within the output destination to store the coded lists\n",
    "    \n",
    "    #Creates temporary variables to manage paths\n",
    "    files = next(os.walk(path_now))[2]  #Helps to check if we are in the main folder generated by Extract_IDRs by compiling a list of all files within this folder\n",
    "    folders = next(os.walk(path_now))[1]   #Helps to make a list of all the groups and theirs paths that we will be going through by compiling a list of all folders within this folder\n",
    "    fastas = {}  #Lists all the fasta files for each group (either .fasta or .txt)\n",
    "\n",
    "    \n",
    "    #Checks if we are in the main folder generated by Extract_IDRs\n",
    "    if 'Summary.txt' in files:\n",
    "        print(\"Extract_IDRs has been used!\"+'\\n')\n",
    "        #Explains where to find output\n",
    "        print(\"Output is set to newly created Slices folder within each group's respective folder insisde Extract_IDRs_Results folder!\" + \"\\n\")\n",
    "        for f in folders:\n",
    "            #Updates folders with the path to the subfolder 'IDRs' where Extract_IDRs stored the fasta sequences of the IDRs\n",
    "            folders[n1] = path_now + folders[n1] +'/IDRs/'\n",
    "            #Compiles the fasta sequences of the IDRs for each group within a variable (either .fasta or .txt)\n",
    "            fastas[n1] = [f for f in os.listdir(folders[n1]) if f.endswith(\".txt\") | f.endswith(\".fasta\")]\n",
    "            n1 = n1 + 1            \n",
    "            \n",
    "        #In case we are in the main folder generated by Extract_IDRs, updates output_path to a list of outputs so that each group folder within\n",
    "        # the Extract_IDRs folder will have their own Slices folder\n",
    "        output_paths = next(os.walk(path_now))[1]   #Helps to make a list of all the groups and theirs paths that we will be going through by compiling a list of all folders within this folder\n",
    "        #Resets n1\n",
    "        n1 = 0        \n",
    "        for f2 in output_paths:\n",
    "            output_paths[n1] = path_now + output_paths[n1]\n",
    "            #Updates output path to aid with folder creation\n",
    "            output_paths[n1] = output_paths[n1]+\"/Slices_of_\"+str(window)+\"_StepSize_\"+str(step)          \n",
    "            n1 = n1 + 1 \n",
    "            \n",
    "\n",
    "    #Checks if we are in a folder containing fasta sequences (either .fasta or .txt), thus we opted not to use Extract_IDRs first\n",
    "    elif files != []:\n",
    "        print(\"Extract_IDRs has not been used!\"+'\\n')\n",
    "        #Explains where to find output\n",
    "        print(\"Output is set to newly created Slices folder in the same destination as input!\" + \"\\n\")\n",
    "        independently = 1 # we are working independently from Extract_IDRs\n",
    "        #Compiles the fasta sequences inside the this folder within a variable (either .fasta or .txt)\n",
    "        fastas[n1] = [f for f in os.listdir(path_now) if f.endswith(\".txt\") | f.endswith(\".fasta\")]       \n",
    "        \n",
    "        #Checks for error: no fasta files inside specified location (neither .fasta nor .txt)\n",
    "        if fastas[n1] == []:\n",
    "            sys.exit(\"No fasta files were found (neither .fasta nor .txt)! \\nPlease, try a different folder! =)\") \n",
    "        \n",
    "        #If there is no error, proceeds with code    \n",
    "        n1 = n1 + 1\n",
    "        \n",
    "        #switches folders to path_now so the program can run independently from Extract_IDRs\n",
    "        if path_now.endswith(\"/\"):\n",
    "            folders = path_now\n",
    "        else:\n",
    "            folders = path_now + \"/\"\n",
    "            \n",
    "        #Defines output path to the same folder where input is, but a folder called slices will be created later\n",
    "        output_path = path_now\n",
    "        #Updates output path to aid with folder creation\n",
    "        output_path = output_path+\"/Slices_of_\"+str(window)+\"_StepSize_\"+str(step)\n",
    "    \n",
    "    \n",
    "    #Exits program because no fasta files were found (either .txt or .fasta)\n",
    "    else:\n",
    "        sys.exit(\"No fasta files were found (neither .fasta nor .txt)! \\nPlease, try a different folder!\") \n",
    "\n",
    "        \n",
    "        \n",
    "    #Preapares to loop through the groups\n",
    "    while n2 < n1:\n",
    "        \n",
    "        if independently == 0:  #Extract IDRs was used\n",
    "            #Changes current directory to the folder containing the fasta sequences (either .fasta or .txt)\n",
    "            os.chdir(folders[n2])\n",
    "            #Prepares to create a folder named 'Slices' to store the sequence slices inside the folder of the current group\n",
    "            os.mkdir(output_paths[n2])\n",
    "            #updates output path to move on with code\n",
    "            output_paths[n2] = output_paths[n2]+\"/\"            \n",
    "            \n",
    "        if independently == 1:  #Extract IDRs was NOT used \n",
    "            #Changes current directory to the folder containing the fasta sequences (either .fasta or .txt)\n",
    "            os.chdir(folders)\n",
    "            #Prepares to create a folder named 'Slices' to store the sequence slices inside the folder of the current group\n",
    "            os.mkdir(output_path)\n",
    "            #updates output path to move on with code\n",
    "            output_path = output_path+\"/\"\n",
    "\n",
    "        \n",
    "        #Starts to loop trhough each fasta file within the group to add them to the list file   \n",
    "        #print(fastas)\n",
    "        counter = 0  #Resets counter\n",
    "        for f in fastas[n2]:\n",
    "            counter = counter + 1\n",
    "            \n",
    "            #Creates handy counters/variables\n",
    "            current_seq_lenth = 0  #for the program to know when to stop\n",
    "            current_position = 0  #current residue / start of the slice\n",
    "            slice_end = 0  #end of the slice residue\n",
    "            n_seq = str(n2)+str(counter)  #keeps track of the current sequence number using group number (n2) and sequence number (counter)\n",
    "            this_name = '' #Creates/resets variable for current protein name\n",
    "            seq = '' #Creates/resets variable for current protein sequence\n",
    "            lines = ''  #Creates/resets variable to store all lines in current fasta file\n",
    "            positions = ''  #Creates/resets variable to store IDR start and end residue number when coming from Extract IDRs\n",
    "                      \n",
    "            \n",
    "            #Loads sequence into variable seq\n",
    "            c = open(f)  #Opens current fasta file\n",
    "            lines = c.readlines()   #Reads all lines into temporary variable lines\n",
    "            for line in lines:\n",
    "                if line.startswith ('>') == True:\n",
    "                    this_name = line   #Extracts protein name into this_name\n",
    "                    #If Extract IDRs was used\n",
    "                    if independently == 0:\n",
    "                        this_name_split = this_name.split(\"|\")   #Splits first line of FASTA file to obtain IDR start and end residue numbers\n",
    "                        positions = this_name_split[-2].split(\"_\")  #Extracts IDR start and end residue numbers\n",
    "                if line.startswith ('>') == False:\n",
    "                    seq = r''.join((seq, line.strip()))  #Extracts protein sequence into seq            \n",
    "            c.close()\n",
    "\n",
    "            #Deletes spaces and \\n from the sequence so it can be properly processed\n",
    "            seq = seq.replace(\" \", \"\")\n",
    "            seq = seq.replace(\"\\n\", \"\")            \n",
    "\n",
    "            #When not using Extract IDRs\n",
    "            if independently == 1:\n",
    "            #Updates current sequence length\n",
    "                current_seq_length = len(seq)\n",
    "            #Resets current position to 0 and updates slice_end\n",
    "                current_position = 0\n",
    "                slice_end = current_position + window\n",
    "            \n",
    "            #When using Extract IDRs\n",
    "            if independently == 0:\n",
    "            #Updates current sequence length\n",
    "                current_seq_length = int(positions[1]) + 1\n",
    "            #Resets current position to 0 and updates slice_end\n",
    "                current_position = int(positions[0])\n",
    "                slice_end = current_position + window\n",
    "\n",
    "\n",
    "            #Starts to loop through the sequence and extract the slices\n",
    "            while slice_end <= current_seq_length:\n",
    "\n",
    "                if independently == 0:  #Extract IDRs was used\n",
    "                    #program will use output_paths[n2]\n",
    "                    \n",
    "                    #Extracts the slice\n",
    "                    current_slice = seq[(current_position-int(positions[0])):(slice_end-int(positions[0]))]\n",
    "                    #print(str(current_position + 1) +\"  \"+ str(current_slice) +\"  \"+ str(slice_end))\n",
    "            \n",
    "                    #For Protein Name, file name: (Preparing to store results)\n",
    "                    protein_name = this_name.strip(\"\\n\")+\"|\"+str(current_position)+\"_\"+str(slice_end-1)\n",
    "                    file_name = f.strip(\".txt\")+\"_\"+str(current_position)+\"_\"+str(slice_end-1)+'.txt'\n",
    "                    \n",
    "                    #Saving Results\n",
    "                    f1 = open(str(output_paths[n2])+str(file_name), \"a\")\n",
    "                    f1.write(str(protein_name)+\"\\n\"+str(current_slice))\n",
    "                    f1.close()\n",
    "            \n",
    "                    #For initial residue, slice sequence, final residue\n",
    "                    #print(str(current_position + 1) +\"  \"+ str(current_slice) +\"  \"+ str(slice_end))\n",
    "\n",
    "                if independently == 1:  #Extract IDRs was NOT used\n",
    "                    #program will use output_path\n",
    "                    \n",
    "                    #Extracts the slice\n",
    "                    current_slice = seq[current_position:slice_end]\n",
    "                    #print(str(current_position + 1) +\"  \"+ str(current_slice) +\"  \"+ str(slice_end))\n",
    "                    \n",
    "                    #For Protein Name, file name: (Preparing to store results)\n",
    "                    protein_name = this_name.strip(\"\\n\")+\"|\"+str(current_position + 1)+\"_\"+str(slice_end)\n",
    "                    file_name = f.strip(\".txt\")+\"_\"+str(current_position + 1)+\"_\"+str(slice_end)+'.txt'\n",
    "                    \n",
    "                    #Saving Results\n",
    "                    f1 = open(str(output_path)+str(file_name), \"a\")\n",
    "                    f1.write(str(protein_name)+\"\\n\"+str(current_slice))\n",
    "                    f1.close()\n",
    "                    \n",
    "                    #For initial residue, slice sequence, final residue\n",
    "                    #print(str(current_position + 1) +\"  \"+ str(current_slice) +\"  \"+ str(slice_end))\n",
    "                \n",
    "                #Moves to the next iteration\n",
    "                current_position = current_position + step\n",
    "                slice_end = slice_end + step\n",
    "       \n",
    "    \n",
    "        #Prepares counter for next iteration\n",
    "        n2 = n2 + 1\n",
    "    \n",
    "    print(\"\\n End \\n\")  \n",
    "    \n",
    "    ##This is what this function returns to be subsequently used in the IDR Sequence Composition Module/Function                                          \n",
    "    return #output_path\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
